// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless  by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// python -m grpc_tools.protoc -I=. --python_out=. --pyi_out=.  --grpc_python_out=. ./chat_app.proto

syntax = "proto3";
import "google/protobuf/timestamp.proto";

option java_multiple_files = true;
option java_package = "io.grpc.examples.PrimaryBackupServer";
option java_outer_classname = "PrimaryBackupServerProto";
option objc_class_prefix = "PBS";

package primarybackupserver;

// Interface exported by the server.
// This service implrements a primary backup server whose state only has users and messages
//
service PrimaryBackupServer {

  //  
  rpc NewUser(User) returns (RequestReply); 

  // 
  rpc DeleteUser(User) returns (RequestReply); 

  // Request enqueue a message.
  rpc EnqueueMessage(Message) returns (RequestReply); 
  
  // Request to dequeue a message.
  rpc DequeueMessage(Message) returns (RequestReply);
}

// message RequestReply
// Standard object for Server message replies indicating the 
// status of the request
message RequestReply {
     string reply = 1; 
     int32 request_status = 2;
}

// message User
// Represents the User object (only username in this case)
message User {
    string username = 1; // Username (uniquely identifies user)
}

// message ChatMessage
// Represent the Message object with the text content and 
// metadata (sender, destinataries and timestamp)
message Message {
    User sender = 1;  // Sender user of the message.
    repeated User destinataries = 2; // Destinatary users of the message.
    string text = 3; // Text body of the message.
    google.protobuf.Timestamp date = 4; // Datetime at which message is sent.
}


